x/staking/keeper/grpc_query.go

```diff
@@ -0,0 +1,445 @@
+		redels, res, err = queryRedelegationsFromSrcValidator(store, k, req)
+	default:
+		redels, res, err = queryAllRedelegations(store, k, req)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	redelResponses, err := RedelegationsToRedelegationResponses(ctx, k.Keeper, redels)
+	if err != nil {
+		return nil, err
+	}
+
+	return &types.QueryRedelegationsResponse{RedelegationResponses: redelResponses, Res: res}, nil
+}
+
+func (k Querier) DelegatorValidators(c context.Context, req *types.QueryDelegatorValidatorsRequest) (*types.QueryDelegatorValidatorsResponse, error) {
+	if req == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "empty request")
+	}
+
+	if req.DelegatorAddr.Empty() {
+		return nil, status.Errorf(codes.InvalidArgument, "invalid request")
+	}
+	ctx := sdk.UnwrapSDKContext(c)
+
+	store := ctx.KVStore(k.storeKey)
+	delStore := prefix.NewStore(store, types.GetDelegationsKey(req.DelegatorAddr))
+
+	var validators types.Validators
+	res, err := query.Paginate(delStore, req.Req, func(key []byte, value []byte) error {
+		delegation, err := types.UnmarshalDelegation(k.cdc, value)
+		if err != nil {
+			return err
+		}
+
+		validator, found := k.GetValidator(ctx, delegation.ValidatorAddress)
+		if !found {
+			return types.ErrNoValidatorFound
+		}
+
+		validators = append(validators, validator)
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &types.QueryDelegatorValidatorsResponse{Validators: validators, Res: res}, nil
+}
+
+// Pool queries the pool info
+func (k Querier) Pool(c context.Context, _ *types.QueryPoolRequest) (*types.QueryPoolResponse, error) {
+	ctx := sdk.UnwrapSDKContext(c)
+	bondDenom := k.BondDenom(ctx)
+	bondedPool := k.GetBondedPool(ctx)
+	notBondedPool := k.GetNotBondedPool(ctx)
+
+	if bondedPool == nil || notBondedPool == nil {
+		return nil, status.Errorf(codes.FailedPrecondition, "pool accounts haven't been set")
+	}
+
+	pool := types.NewPool(
+		k.bankKeeper.GetBalance(ctx, notBondedPool.GetAddress(), bondDenom).Amount,
+		k.bankKeeper.GetBalance(ctx, bondedPool.GetAddress(), bondDenom).Amount,
+	)
+
+	return &types.QueryPoolResponse{Pool: pool}, nil
+}
+
+// Params queries the staking parameters
+func (k Querier) Params(c context.Context, _ *types.QueryParamsRequest) (*types.QueryParamsResponse, error) {
+	ctx := sdk.UnwrapSDKContext(c)
+	params := k.GetParams(ctx)
+
+	return &types.QueryParamsResponse{Params: params}, nil
+}
+
+func queryRedelegation(store sdk.KVStore, k Querier, req *types.QueryRedelegationsRequest) (redels types.Redelegations, res *query.PageResponse, err error) {
+	redStore := prefix.NewStore(store, types.GetREDKey(req.DelegatorAddr, req.SrcValidatorAddr, req.DstValidatorAddr))
+
+	var redel types.Redelegation
+	res, err = query.Paginate(redStore, req.Req, func(key []byte, value []byte) error {
+		redel, err = types.UnmarshalRED(k.cdc, value)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
```

Comment:
Do we need paginate here? Paginate logic looks incorrect. It will return only the last value from paginate. Can you add a test with more than 100 entries?

Reply:
Yeah, pagination doesn't have an effect here as it returns only one pair in this case. It's needed for other two cases though